### PR TITLE
Subclass API (#966)

### PR DESCRIPTION
**Pull request by metascroy**
Summary:

Adds new int8_dynamic_activation_intx_weight quantization with subclass API

Differential Revision: D62464487

### CODE REVIEW

File: torchao/quantization/quant_primitives.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def _quantize_affine_no_dtype_cast(
     elif zero_point_domain is None:
         # This case handles quantization for float8 we expect no zero point and no zero point domain
         assert zero_point is None, "zero_point should be None when zero_point_domain is None"
-        quant = torch.clamp(input * scale.reciprocal(), quant_min, quant_max)
+        quant = torch.clamp(torch.round(input * (1.0 / scale)), quant_min, quant_max)
```
Comment (metascroy) [inline torchao/quantization/quant_primitives.py:303]: @jerryzh168 to confirm if this is OK.  It was needed to match behavior of other quantizer.

File: torchao/quantization/quant_primitives.py [hunk 2]
```diff
@@ -766,13 +766,16 @@ def _choose_qparams_affine(
         assert mapping_type == MappingType.ASYMMETRIC.name
         scale = (max_val_pos - min_val_neg) / float(quant_max - quant_min)
         scale = torch.clamp(scale, min=eps)
-        if preserve_zero:
-            zero_point = quant_min - torch.round(min_val_neg / scale)
-            zero_point = torch.clamp(zero_point, quant_min, quant_max)
+        if zero_point_domain is None:
```
Comment (metascroy) [inline torchao/quantization/quant_primitives.py:769]: @jerryzh168 confirm if this is OK.  It was needed to get scale-only quantization in affine_quantized_tensor

File: torchao/experimental/tests/test_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        )
+
+        unwrapped_model = copy.deepcopy(model)
+        unwrap_tensor_subclass(model)
+
+        print("Exporting quantized model")
+        exported = torch.export.export(model, (activations,))
+        
+        print("Compiling quantized model")
+        compiled = torch.compile(unwrapped_model)
```
Comment (metascroy) [inline torchao/experimental/tests/test_linear_8bit_act_xbit_weight_subclass_quantizer.py:135]: @jerryzh168 do you see unification for compile and export coming soon?  The fact that one requires an unwrapped tensor subclass and the other requires a wrapped one makes using this API inconvenient in torchchat.

**Comment by metascroy** [comment]
@kimishpatel @jerryzh168 moving review over to GH.  I hope I've addressed most of your concerns.

@jerryzh168, the fact that compile and export cannot handle the same model (export requires an unwrapped tensor subclass, compile requires a wrapped one, and eager can handle both) makes using this API inconvenient in torchchat.  Do you know if there is planned unification there?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    assert input_tensor.dim() >= 3
+    lead_shape = input_tensor.shape[0:-2]
+    m, k = input_tensor.shape[-2], input_tensor.shape[-1]
+    n, k_ = weight_tensor.shape
+    assert k_ == k
+
+    input_tensor = input_tensor.reshape(-1, m, k)
+
+    res = [
+        _impl_2d(input_tensor[i, :, :], weight_tensor)
```
Comment (kimishpatel) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:382]: Why are you doing it like this? You can just fuse first N dim. LIke line 379 should be
`input_tensor = input_tensor.reshape(-1, k)` no?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        m, k = input_tensor.shape
+        n, k_ = weight_tensor.shape
+        assert k_ == k
+        group_size = weight_tensor.layout_tensor.layout_type.group_size
+        packed_weight = weight_tensor.layout_tensor.packed_weight
+
+        # TODO(T200095131): convert self.n, self.k, self.group_size to
+        # int when supported by AOTI
+        args = (
+            input_tensor,
```
Comment (kimishpatel) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:345]: Does this expect input_tensor to be of fp32 type? Would it work with bf16? If not we should assert for those

Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:345]: I can add an assert here too, although this is already asserted in the C++ layer

File: torchao/quantization/quant_primitives.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:303]: hmmm, it might be fine as long as all the tests passes I think

File: torchao/quantization/quant_primitives.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:769]: OK, should zero_point be `None` here?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+            ),
+            quant_min=None,  # TODO: why is this an arg for this function
+            quant_max=None,  # TODO: why is this an arg for this function
+            dtype=None,  # TODO: why is this an arg for this function
+            group_size=group_size,
+            output_dtype=torch.float32,
+        )
+
+        # Quantize activations
+        activation_scales, activation_zeros = choose_qparams_affine(
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:269]: dynamic quantization should be reusing affine quantized tensor, example: https://github.com/pytorch/ao/blob/900f9acccb8bc3c102d3a7095d1b1f0355057bec/torchao/quantization/quant_api.py#L586

why is this calling these functions here?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+        weight_qvals = weight_tensor.layout_tensor.packed_weight.to(torch.int32)
+        weight_scales = weight_tensor.layout_tensor.scale
+        weight_zeros = weight_tensor.layout_tensor.zero_point
+        group_size = weight_tensor.layout_tensor.layout_type.group_size
+        has_weight_zeros = weight_zeros is not None
+        m, k = input_tensor.shape
+        n, k_ = weight_tensor.shape
+        assert k_ == k
+
+        weights_dequantized = dequantize_per_channel_group(
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:255]: we tend to use quantize_affine/dequantize_affine I think, also this should probably be:

`weights_dequantized = weight_tensor.dequantize()`?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+
+
+def _linear_int8_dynamic_activation_intx_weight_check(
+    input_tensor, weight_tensor, bias
+):
+    layout_type = weight_tensor.layout_tensor.layout_type
+    return isinstance(layout_type, IntxWeightLayoutType) and bias is None
+
+
+def _linear_int8_dynamic_activation_intx_weight_fallback_impl(
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:236]: see fallback here: https://github.com/pytorch/ao/blob/900f9acccb8bc3c102d3a7095d1b1f0355057bec/torchao/dtypes/affine_quantized_tensor.py#L1760-L1764 I think you should be able to do the same here

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+
+
+def _target_equal(target, other):
+    assert other in _VALID_TARGETS
+    assert target in _VALID_TARGETS
+    return target == other
+
+
+# This format is intended for use with int8 dynamic quantization
+class IntxWeightLayoutType(LayoutType):
```
[the remark this answers is not in the document]
Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:51]: This layout targets the linear_8bit_act_xbit_weight kernels in torchao/experimental.  They can run on all PyTorch platforms (eager, AOTI, compile, and ExecuTorch), so adding ExecuTorch doesn't make sense.  Open to naming suggestions.

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 6 above)
Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:269]: That function doesn't look equivalent?  It looks like the quantization is symmetric.

File: torchao/quantization/quant_primitives.py (hunk 1 above)
Comment (metascroy) [reply torchao/quantization/quant_primitives.py:303]: I think the tests do not pass because it is slightly different quantization logic.  It looks more sensible to me to round before truncating, but I can also drop this change.

We can do a perplexity study when moving from the other quantizer to this one in torchchat.  But I have narrowed done this as being the only numerical difference between the two.

File: torchao/quantization/quant_primitives.py (hunk 2 above)
Comment (metascroy) [reply torchao/quantization/quant_primitives.py:769]: I could make it None, but that changes the return type of this method from Tuple[Tensor, Tensor] to Tuple[Tensor, Optional[Tensor]]

**Comment by kimishpatel** [comment]
I plan to review some time tomorrow

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 8 above)
Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:241]: Used weight_tensor.dequantize() below

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 5 above)
Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:334]: Actually, I like keeping the assert in the C++ layer.  I think it makes sense for the op to assert its inputs rather than the thing calling the op to assert the inputs.

File: torchao/experimental/quant_api.py [hunk 10]
```diff
@@ -317,3 +317,53 @@ def quantize(self, model: nn.Module) -> nn.Module:
...
+def int8_dyn_act_intx_weight(
+    group_size: int = 128,
+    nbit: int = 4,
+    has_weight_zeros: bool = False,
+    target: str = "native",
+):
+
+    def apply(weight):
+        assert weight.shape[-1] % group_size == 0
+        assert weight.device == torch.device("cpu"), "Only CPU is supported"
```
Comment (metascroy) [inline torchao/experimental/quant_api.py:340]: Add CPU device assert here

File: torchao/quantization/quant_primitives.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:304]: I see, we don't want to break tests I think, but if this is better for torchchat we can create a new quant primitive op or add a new option here I feel

File: torchao/quantization/quant_primitives.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:769]: yeah I think making it None probably makes more sense here

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 6 above)
Comment (jerryzh168) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:263]: you can choose a different mapping type,
I mean we can use https://github.com/pytorch/ao/blob/900f9acccb8bc3c102d3a7095d1b1f0355057bec/torchao/quantization/quant_api.py#L613 to do activation quantization, maybe "Dynamic Activation Quantization + Weight Quantization" in https://github.com/pytorch/ao/issues/391 is clearer in explaining this, please let me know if it makes sense

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:51]: OK I feel it makes sense to just include the kernel name here then, like `Linear8BitActXBitWeightLayoutType` (also we are renaming LayoutType to Layout to make things clearer) you'll probably see this after rebase

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 6 above)
Comment (metascroy) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:263]: Sorry, I don't follow what you're suggesting.  I have weights_dequantized on line 260, and I need activations_dequantized so I can call torch.matmul(activations_dequantized, weights_dequantized.transpose(1, 0)) on line 296.

I'm not sure what I'm suppose to replace the code that generates activations_dequantized with.

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
some requested changes, please see comments

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py]: @kimishpatel I don't quite follow why "it's not feasible to enumerate all possible ways in which weights can be packed" but I think we are not asking people to use AffineQuantizedTensor at this stage as I discussed in the post, so feel free to contribute in a way you feel makes more sense, we can always merge/refactor later if needed. although a high level API + brief description of implementation might be helpful for us to understand what you have in mind.

also for "Quant API shouldnt really be in the business of understanding packed layout information. This should really be left to subclasses of AQT" I think it might be better/easier to just copy paste AQT and create a new tensor subclass in that case instead of subclassing AQT, we are not clear whether we should have inheriting AQT as an extension point yet

File: torchao/quantization/quant_primitives.py [hunk 11]
```diff
@@ -344,6 +346,8 @@ def _quantize_affine_no_dtype_cast(
         quant = torch.clamp(
             torch.round(input * (1.0 / scale)) + zero_point, quant_min, quant_max
         )
+    elif zero_point_domain == ZeroPointDomain.ZERO.name:
+        quant = torch.clamp(torch.round(input * (1.0 / scale)), quant_min, quant_max)
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:350]: nit: maybe add an assert here to check zero_point is None

File: torchao/experimental/docs/readme.md
Comment (jerryzh168) [inline torchao/experimental/docs/readme.md:1]: readme.md --> README.md?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:1]: nit: I think we should remove the quantizer in the name, `_linear_8bit_act_xbit_weight_layout.py` might be more appropriate

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        group_size = weight_tensor.tensor_impl.get_layout().group_size
+        has_weight_zeros = weight_zeros is not None
+        m, k = input_tensor.shape
+        n, k_ = weight_tensor.shape
+        assert k_ == k
+
+        weights_dequantized = weight_tensor.dequantize()
+
+        # Quantize activations
+        activation_scales, activation_zeros = choose_qparams_affine(
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:269]: I think we can probably use `to_affine_quantized_intx` and `dequantize_affine` here to quantize activation?

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+from enum import auto, Enum
+
+import logging
+from typing import List, Optional, Tuple
+
+import torch
+from torch.ao.quantization.fx._decomposed import (
+    dequantize_per_channel_group,
+    quantize_per_channel_group,
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:15]: these ops are a bit deprecated so we tend not to use these if possible

File: torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+    register_aqt_quantized_linear_dispatch,
+    register_layout,
+)
+from torchao.dtypes.utils import Layout
+from torchao.quantization.quant_primitives import (
+    choose_qparams_affine,
+    MappingType,
+    ZeroPointDomain,
+)
+from torchao.utils import TorchAOBaseTensor
```
Comment (jerryzh168) [inline torchao/experimental/_linear_8bit_act_xbit_weight_subclass_quantizer.py:29]: nit: looks like not used

**Review by jerryzh168 — APPROVED** [review]
LGTM, left a few more nit comments

File: torchao/quantization/quant_primitives.py [hunk 15]
```diff
@@ -60,9 +60,11 @@ class ZeroPointDomain(Enum):
 
     integer domain: quantized_val = (float_val / scale) (integer) + zero_point (integer)
     float domain: quantized_val = (float_val - (zero_point (float) - scale * mid_point)) / scale
+    zero domain: quantized_val = (float_val / scale)
     """
     INT = auto()
     FLOAT = auto()
+    ZERO = auto()
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:67]: btw, I feel maybe NONE would be better?

File: torchao/experimental/quant_api.py [hunk 16]
```diff
@@ -463,3 +463,56 @@ def quantize(self, model: nn.Module) -> nn.Module:
...
+from torchao.experimental._linear_8bit_act_xbit_weight_layout import Linear8BitActXBitWeightLayout
+from torchao.quantization.quant_api import (
+    _get_linear_subclass_inserter,
+    MappingType,
+    to_affine_quantized_intx,
+    ZeroPointDomain,
+)
+
+
+def int8_dyn_act_intx_weight(
```
Comment (jerryzh168) [inline torchao/experimental/quant_api.py:477]: we are spelling out all the words in our API so far, so this should be `int8_dynamic_activation_intx_weight` I think

File: torchao/experimental/tests/test_linear_8bit_act_xbit_weight_subclass_quantizer.py
Comment (jerryzh168) [inline torchao/experimental/tests/test_linear_8bit_act_xbit_weight_subclass_quantizer.py:1]: can you rename the tests as well

**Review by jerryzh168 — APPROVED** [review]
